File: JavaScript/leetcode75/src/main/Problem36_PathSumIII.js
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
var pathSum = function (root, sum) {
  if (root == null) return 0;
  return pathSum(root.left, sum) + pathSum(root.right, sum) + helper(root, sum);
};

function helper(root, sum, count) {
  if (root == null) return 0;
  var count = 0;
  if (sum == root.val) {
    count++;
  }
  count += helper(root.left, sum - root.val);
  count += helper(root.right, sum - root.val);
  return count;
}

module.exports = pathSum;

File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem36_PathSumIII.java
package org.phoenix.leetcode.challenges;

import org.phoenix.leetcode.challenges.Problem33_MaximumDepthOfBinaryTree.TreeNode;

public class Problem36_PathSumIII {
    int res = 0;

    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) return 0;

        helper(root, targetSum);
        pathSum(root.left, targetSum);
        pathSum(root.right, targetSum);
        return res;
    }

    public void helper(TreeNode root, long sum) {
        if (root == null) {
            return;
        }
        if (sum - root.val == 0) {
            res++;
        }
        helper(root.left, sum - root.val);
        helper(root.right, sum - root.val);
    }
}
